alter table add index reports 'panic in the recoverable goroutine'

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
drop table if exists t;
set @@tidb_enable_clustered_index=true;
create table t (a int, b enum('Alice'), c int, primary key (c, b));
insert into t values (-1,'Alice',100);
insert into t values (-1,'Alice',7000);
split table t between (0,'Alice') and (10000,'Alice') regions 2;
alter table t add index idx (c);
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
```console
Query OK, 0 rows affected (2.53 sec)
```

### 3. What did you see instead (Required)
```
[ERROR] [misc.go:115] ["panic in the recoverable goroutine"] [label=ddl] [funcInfo=backfillWorker.run] [r={}] [stack="goroutine 696 [running]:\ngithub.com/pingcap/tidb/util.GetStack(...)\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/util/misc.go:76\ngithub.com/pingcap/tidb/util.Recover(0x36eae97, 0x3, 0x370ea96, 0x12, 0x0, 0x0)\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/util/misc.go:119 +0x328\npanic(0x3513a60, 0xc000cbc400)\n\t/usr/local/go/src/runtime/panic.go:969 +0x1b9\ngithub.com/pingcap/tidb/util/codec.CutOne(0xc00fabfb14, 0x6, 0x10, 0x0, 0x1, 0xc01068e340, 0x0, 0x4, 0x10, 0x0, ...)\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/util/codec/codec.go:895 +0x125\ngithub.com/pingcap/tidb/kv.NewCommonHandle(0xc00fabfb0b, 0xf, 0x19, 0x46fb8d0, 0x31, 0x0)\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/kv/key.go:248 +0xd4\ngithub.com/pingcap/tidb/tablecodec.DecodeRowKey(0xc00fabfb00, 0x1a, 0x24, 0x0, 0x0, 0x1, 0x372a4a4)\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/tablecodec/tablecodec.go:269 +0x225\ngithub.com/pingcap/tidb/ddl.tryDecodeToHandleString(0xc00fabfb00, 0x1a, 0x24, 0x2, 0xc000ce0c38)\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/ddl/backfilling.go:422 +0x49\ngithub.com/pingcap/tidb/ddl.(*reorgBackfillTask).String(0xc00f8d34c0, 0xc010019d18, 0x1)\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/ddl/backfilling.go:185 +0x87\ngithub.com/pingcap/tidb/ddl.(*backfillWorker).run(0xc000cc39e0, 0xc000ccc0b0, 0x3b01ec0, 0xc0105c3e10)\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/ddl/backfilling.go:299 +0x467\ncreated by github.com/pingcap/tidb/ddl.(*worker).writePhysicalTableRecord\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/ddl/backfilling.go:598 +0xa2a\n"]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
master `8ddd41c960caaebbdeb28da33c781fca1464f05f`

--- a/executor/clustered_index_test.go
+++ b/executor/clustered_index_test.go
@@ -274,3 +274,16 @@ func (s *testClusteredSuite) TestClusteredIndexSplitAndAddIndex(c *C) {
 	tk.MustQuery("select a from t order by a;").Check(testkit.Rows("a", "b", "c", "u"))
 	tk.MustQuery("select a from t use index (idx) order by a;").Check(testkit.Rows("a", "b", "c", "u"))
 }
+
+// https://github.com/pingcap/tidb/issues/22453
+func (s *testClusteredSerialSuite) TestClusteredIndexSplitAndAddIndex2(c *C) {
+	tk := s.newTK(c)
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (a int, b enum('Alice'), c int, primary key (c, b));")
+	tk.MustExec("insert into t values (-1,'Alice',100);")
+	tk.MustExec("insert into t values (-1,'Alice',7000);")
+	tk.MustQuery("split table t between (0,'Alice') and (10000,'Alice') regions 2;").Check(testkit.Rows("1 1"))
+	tk.MustExec("set @@global.tidb_ddl_error_count_limit = 3;")
+	tk.MustExec("alter table t add index idx (c);")
+	tk.MustExec("admin check table t;")
+}

--- a/util/codec/codec.go
+++ b/util/codec/codec.go
@@ -921,6 +921,7 @@ func SetRawValues(data []byte, values []types.Datum) error {
 
 // peek peeks the first encoded value from b and returns its length.
 func peek(b []byte) (length int, err error) {
+	originLength := len(b)
 	if len(b) < 1 {
 		return 0, errors.New("invalid encoded key")
 	}
@@ -952,6 +953,10 @@ func peek(b []byte) (length int, err error) {
 		return 0, errors.Trace(err)
 	}
 	length += l
+	if length > originLength {
+		return 0, errors.Errorf("invalid encoded key, " +
+			"expected length: %d, actual length: %d", length, originLength)
+	}
 	return
 }
 

--- a/util/codec/codec.go
+++ b/util/codec/codec.go
@@ -954,7 +954,7 @@ func peek(b []byte) (length int, err error) {
 	}
 	length += l
 	if length > originLength {
-		return 0, errors.Errorf("invalid encoded key, " +
+		return 0, errors.Errorf("invalid encoded key, "+
 			"expected length: %d, actual length: %d", length, originLength)
 	}
 	return

--- a/util/codec/codec_test.go
+++ b/util/codec/codec_test.go
@@ -927,6 +927,15 @@ func (s *testCodecSuite) TestCut(c *C) {
 	c.Assert(n, Equals, int64(42))
 }
 
+func (s *testCodecSuite) TestCutOneError(c *C) {
+	var b []byte
+	_, _, err := CutOne(b)
+	c.Assert(err, ErrorMatches, "invalid encoded key")
+	b = []byte{4 /* codec.uintFlag */, 0, 0, 0}
+	_, _, err = CutOne(b)
+	c.Assert(err, ErrorMatches, "invalid encoded key.*")
+}
+
 func (s *testCodecSuite) TestSetRawValues(c *C) {
 	sc := &stmtctx.StatementContext{TimeZone: time.Local}
 	datums := types.MakeDatums(1, "abc", 1.1, []byte("def"))

